modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp

```diff
@@ -0,0 +1,204 @@
+        return;
+
+    size = pInfoBM->bmiHeader.biSizeImage;
+    data.Allocate(size);
+    GetDIBits(hdc, hbm, 0, pInfoBM->bmiHeader.biHeight, data, pInfoBM, DIB_RGB_COLORS);
+
+    SelectObject(hdc, obj);
+    DeleteDC(hdc);
+}
+
+static bool GetIconData(HICON icon, CComHeapPtr<BYTE>& colorData, DWORD& colorSize, CComHeapPtr<BYTE>& maskData, DWORD& maskSize)
+{
+    ICONINFO iconinfo;
+
+    if (!GetIconInfo(icon, &iconinfo))
+        return false;
+
+    ExtractOneBitmap(iconinfo.hbmColor, colorData, colorSize);
+    ExtractOneBitmap(iconinfo.hbmMask, maskData, maskSize);
+
+    DeleteObject(iconinfo.hbmColor);
+    DeleteObject(iconinfo.hbmMask);
+
+    return true;
+}
+
+
+START_TEST(SHCreateFileExtractIconW)
+{
+    WCHAR CurrentModule[MAX_PATH];
+    HMODULE shell32 = LoadLibraryA("shell32.dll");
+    HICON myIcon;
+    pSHCreateFileExtractIconW = (HRESULT (__stdcall *)(LPCWSTR, DWORD, REFIID, void **))GetProcAddress(shell32, "SHCreateFileExtractIconW");
+
+    CoInitialize(NULL);
+
+    GetModuleFileNameW(NULL, CurrentModule, _countof(CurrentModule));
+    {
+        SHFILEINFOW shfi;
+        ULONG_PTR firet = SHGetFileInfoW(CurrentModule, 0, &shfi, sizeof(shfi), SHGFI_ICON);
+        myIcon = shfi.hIcon;
+        if (!firet)
+        {
+            skip("Unable to get my own icon\n");
+            return;
+        }
+    }
+
+    if (!pSHCreateFileExtractIconW)
+    {
+        skip("SHCreateFileExtractIconW not available\n");
+        return;
+    }
+
+    for (size_t n = 0; n < _countof(Tests); ++n)
+    {
+        test_data& cur = Tests[n];
+        bool useMyIcon = false;
+
+        if (cur.Name == NULL)
+        {
+            cur.Name = CurrentModule;
+            useMyIcon = true;
+        }
+
+        CComPtr<IExtractIconW> spExtract;
+        HRESULT hr = pSHCreateFileExtractIconW(cur.Name, cur.dwFlags, IID_PPV_ARG(IExtractIconW, &spExtract));
+        ok(hr == S_OK, "Expected hr to be S_OK, was 0x%lx for %S(%lx)\n", hr, cur.Name, cur.dwFlags);
+
+        if (!SUCCEEDED(hr))
+            continue;
+
+        int ilIndex = -1;
+        UINT wFlags = 0xdeaddead;
+        WCHAR Buffer[MAX_PATH];
+
+        hr = spExtract->GetIconLocation(0, Buffer, _countof(Buffer), &ilIndex, &wFlags);
+        ok(hr == S_OK, "Expected hr to be S_OK, was 0x%lx for %S(%lx)\n", hr, cur.Name, cur.dwFlags);
+        if (!SUCCEEDED(hr))
+            continue;
+
+        ok(wFlags & (GIL_NOTFILENAME|GIL_PERCLASS), "Expected GIL_NOTFILENAME|GIL_PERCLASS to be set for %S(%lx)\n", cur.Name, cur.dwFlags);
+        ok(!wcscmp(Buffer, L"*"), "Expected '*', was '%S' for %S(%lx)\n", Buffer, cur.Name, cur.dwFlags);
+
+        HICON ico;
+        hr = spExtract->Extract(Buffer, ilIndex, &ico, NULL, 0);
+
+        /* Visualize the icon extracted for whoever is stepping through this code. */
+        HWND console = GetConsoleWindow();
+        SendMessage(console, WM_SETICON, ICON_BIG, (LPARAM)ico);
```

Comment:
I think you're supposed to either destroy or restore the old icon returned from this call.